### ISSUE TITLE
Guide to precompile assets in local environment before deploying to Heroku

### ISSUE THREAD

**Issue by woniesong92** [rails/webpacker#1319]
Environment: Rails 5.1.1, webpacker 2.0.

I noticed it takes too long to compile on the fly when deploying to Heroku (~30 mins), and decided to give precompiling locally a shot. I run the following command to precompile:

```
RAILS_ENV=production NODE_ENV=production SECRET_KEY_BASE=.. bundle exec rake assets:clean assets:precompile
```

However, I run into the following problem:

`Module build failed: Error: Couldn't find preset "stage-0" relative to directory "my/app/root/path"`

I made sure to try both `yarn add babel-preset-stage-0` and `npm install --save-dev babel-preset-stage-0`. What am I doing wrong?

**Comment by woniesong92** [comment]
Ah, I got it working by adding a few more environment variables. It wasn't working because `babel-preset-stage-0` was in `devDependencies`.

```
RAILS_ENV=production NODE_ENV=production NPM_CONFIG_PRODUCTION=false YARN_PRODUCTION=false SECRET_KEY_BASE=... bundle exec rake assets:clean assets:precompile
```

Then, to let Heroku know that I've precompiled assets locally, I check in `public/assets` folder.

```
git add public assets && git commit -m 'check in assets'
```

Am I also expected to include the files in `public/packs`?

**Comment by woniesong92** [comment]
I solved the issue, but not sure if it's just a bypass or an actual solution.

TLDR: commented out uglifying in production config, and it compiled fast enough.

This helped a lot: https://github.com/rails/webpacker/issues/1217#issuecomment-361425935

This solution is good enough for me now, but it's probably not ideal for a lot of other people. Any suggestion?

Other things I tried:
- Upgrade to latest webpacker (3.3.0)
- Run Nodejs buildpack before Ruby buildpack (specific to heroku)
- Get rid of unused libraries from package.json
- Try to precompile locally and check in public/assets (always compiled successfully, but there were always path-related problems on prod)

**Comment by woniesong92** [comment]
For now, I resorted to installing the latest `uglifyjs-webpack-plugin` (not the one that comes with webpacker) and using the following options:

```
    new UglifyJsPlugin({
      sourceMap: false, // sourceMap slows it down even more!
      parallel: true,
    }),
```

Please let me know if you have other suggestions.

**Comment by sunnyrjuneja** [comment]
I have an **extremely** simple webpacker setup (just importing fontawesome 5) and it still takes over 5 minutes. I can't help but feel like something is misconfigured / can be fixed that would make it so that we don't need to precompile the assets. I even have nodejs buildpack before the Ruby buildpack.

**Comment by gauravtiwari** [comment]
@woniesong92 @sunnyrjuneja How is your project structured? Do you have all files under packs/ directory?

**Comment by sunnyrjuneja** [comment]
Hey @gauravtiwari,

Thanks for responding. Here is the is my application.js.

```
import fontawesome from '@fortawesome/fontawesome'
import solid from '@fortawesome/fontawesome-pro-solid'
import regular from '@fortawesome/fontawesome-pro-regular'
import light from '@fortawesome/fontawesome-pro-light'
import brands from '@fortawesome/fontawesome-free-brands'

import MyTypeScriptClass from './my_typescript_class.ts'
window.MyTypescriptClass = MyTypeScriptClass

$(document).on("click", ".specific-class", function(event) {
  // my js
});

fontawesome.library.add(solid, regular, light, brands)
```

Font Awesome 5 lives in node_modules and my directory looks like

```
app/javascript/packs
├── application.js
└── my_typescript_class.ts
```

**Comment by gauravtiwari** [comment]
@sunnyrjuneja Thanks. I see that `my_typescript_class.ts` is inside packs folder and also you have imported it to application.js as well (causing circular dependency). 

Is `my_typescript_class.ts` an entry file (i.e. is it used directly inside the views)? From what I see it seems like `application.js` is the main entry file, right? Please move `my_typescript_class.ts` to `app/javascript` folder. May be something like: 

```yml
app/javascript: 
  src: 
    - my_typescript_class.ts
    - all src files (i.e. not webpack entry files)
  packs:
    # import src files and bootstrap
    - all entry files (i.e. you are going to include in the view)
```
https://webpack.js.org/concepts/entry-points/

**Comment by sunnyrjuneja** [comment]
@gauravtiwari Hey Gaurav,

Thanks for responding. I don't know much about webpack so I appreciate you explaining this to me. 

I'm not sure if I completely understand what you mean by "directly in the views." Right now, MyTypescriptClass is being used in click events and returned in by controller action to trigger changes in the view (rails ujs). I think this qualifies as being used in directly in the views :).

Based on your description, this is what I did. I moved over my_typescript_class.ts to app/javascript/src (from app/javascript/packs).

```
app/javascript
├── packs
│   └── application.js
└── src
    └── my_typescript_class.ts
```

I updated my application.js to reflect the new path:

```
-import EditVacantRow from './my_typescript_class'
+import EditVacantRow from 'src/my_typescript_class'
window.MyTypescriptClass = MyTypescriptClass

$(document).on("click", ".specific-class", function(event) {
  // my js
});
```

(In case it wasn't clear, I use MyTypescriptClass where the click events on `.specific-class`.)

Unfortunately, this didn't help. When I deploy to Heroku, I still see long compilation times.

```
remote:        Compiling…
remote:        Compiled all packs in /tmp/build_6ba3590aa0151c1904847c41fbad9dc1/public/packs
remote:        Asset precompilation completed (394.36s)
```

Did I misunderstand you? Would it help if I created an open source repo to reproduce this?
